cmd/cpe/generate.go

```diff
@@ -2,80 +2,152 @@ package main
 
 import (
 	"compress/gzip"
-	"flag"
+	"crypto/sha256"
 	"fmt"
+	"github.com/facebookincubator/nvdtools/cpedict"
+	"github.com/facebookincubator/nvdtools/wfn"
+	"github.com/fleetdm/fleet/v4/pkg/fleethttp"
+	"github.com/fleetdm/fleet/v4/server/ptr"
+	"github.com/fleetdm/fleet/v4/server/vulnerabilities/nvd"
+	"github.com/pandatix/nvdapi/common"
+	"github.com/pandatix/nvdapi/v2"
 	"io"
-	"net/http"
+	"log"
+	"log/slog"
 	"os"
 	"path/filepath"
 	"strings"
+	"time"
+)
 
-	"github.com/facebookincubator/nvdtools/cpedict"
-	"github.com/fleetdm/fleet/v4/server/vulnerabilities/nvd"
+const (
+	httpClientTimeout = 2 * time.Minute
+	waitTimeForRetry  = 30 * time.Second
+	maxRetryAttempts  = 10
+	apiKeyEnvVar      = "NVD_API_KEY" //nolint:gosec
 )
 
-func panicif(err error) {
+func panicIf(err error) {
 	if err != nil {
 		panic(err)
 	}
 }
 
 func main() {
-	var verbose bool
-	flag.BoolVar(&verbose, "verbose", false, "Sets verbose mode")
-	flag.Parse()
+	apiKey := os.Getenv(apiKeyEnvVar)
+
+	logHandler := slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{Level: slog.LevelInfo})
+	slog.SetDefault(slog.New(logHandler))
+
+	if apiKey == "" {
+		log.Fatal(fmt.Sprintf("Must set %v environment variable", apiKeyEnvVar))
+	}
+
+	cwd, err := os.Getwd()
+	panicIf(err)
+	slog.Info(fmt.Sprintf("CWD: %v", cwd))
 
-	dbPath := cpe()
+	client := fleethttp.NewClient(fleethttp.WithTimeout(httpClientTimeout))
+	dbPath := getCPEs(client, apiKey, cwd)
 
-	fmt.Printf("Sqlite file %s size: %.2f MB\n", dbPath, getSizeMB(dbPath))
+	slog.Info(fmt.Sprintf("Sqlite file %s size: %.2f MB\n", dbPath, getSizeMB(dbPath)))
 
-	fmt.Println("Compressing DB...")
+	slog.Info("Compressing DB...")
 	compressedPath, err := compress(dbPath)
-	panicif(err)
+	panicIf(err)
 
-	fmt.Printf("Final compressed file %s size: %.2f MB\n", compressedPath, getSizeMB(compressedPath))
-	fmt.Println("Done.")
+	slog.Info("Calculating SHA256...")
+	compressedPath, err = addSHA256(compressedPath)
+	panicIf(err)
+
+	slog.Info(fmt.Sprintf("Final compressed file %s size: %.2f MB\n", compressedPath, getSizeMB(compressedPath)))
+	slog.Info("Done.")
 }
 
 func getSizeMB(path string) float64 {
 	info, err := os.Stat(path)
-	panicif(err)
+	panicIf(err)
 	return float64(info.Size()) / 1024.0 / 1024.0
 }
 
-func cpe() string {
-	fmt.Println("Starting CPE sqlite generation...")
+func getCPEs(client common.HTTPClient, apiKey string, resultPath string) string {
+	slog.Info("Fetching CPEs from NVD...")
+
+	nvdClient, err := nvdapi.NewNVDClient(client, apiKey)
+	panicIf(err)
+
+	var cpes []cpedict.CPEItem
+	retryAttempts := 0
+
+	totalResults := 1
+	for startIndex := 0; startIndex < totalResults; {
+		cpeResponse, err := nvdapi.GetCPEs(nvdClient, nvdapi.GetCPEsParams{StartIndex: ptr.Int(startIndex)})
```

Comment:
Their recommendation is to sleep 6 seconds between requests:
![Screenshot 2023-11-10 at 09 44 53](https://github.com/fleetdm/fleet/assets/2073526/83528884-3b0f-4dfb-9714-3b3a75747deb)

Source: https://nvd.nist.gov/developers/api-workflows